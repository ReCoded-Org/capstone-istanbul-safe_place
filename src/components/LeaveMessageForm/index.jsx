import React from "react";
import { Col } from "react-bootstrap";
import "./index.scss";

export default function LeaveMessageForm() {
  return (
    <Col md={6} xs={12}>
      as
    </Col>
  );
}
