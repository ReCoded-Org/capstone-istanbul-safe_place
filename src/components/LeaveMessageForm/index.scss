@import "../../style/variables";
@import "../../style/buttons";
@import "../../style/forms";
@import "../../style/bootstrapImports";

.contactForm {
  font: inherit;

  input[type="text"],
  input[type="email"],
  textarea {
    @include inputElement;
  }

  textarea {
    height: 8rem;
  }

  input[type="submit"] {
    @include primaryBtn;
    background-color: $successColor;
    font-weight: 800;

    &:hover {
      background-color: $lightColor;
      color: $primaryColor;
    }
  }

  .formInput {
    margin-bottom: 1.4rem;
    position: relative;

    label {
      width: 100%;
      font-weight: 500;
    }
  }

  input.inputError,
  textarea.inputError {
    @include inputErrorBorder;
  }
}

@include media-breakpoint-down(sm) {
  .contactForm {
    input[type="submit"] {
      width: 100%;
    }
  }
}
