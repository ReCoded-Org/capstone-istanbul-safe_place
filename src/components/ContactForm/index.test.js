import React from "react";
import renderer from "react-test-renderer";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import LeaveMessageForm from "./index";

describe("LeaveMessageForm", () => {
  it("renders correctly and matches the snapshot", () => {
    const tree = renderer.create(<LeaveMessageForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  const mockSubmit = jest.fn((name, email, messageContent) => {
    return Promise.resolve({ name, email, messageContent });
  });

  beforeEach(() => {
    render(<LeaveMessageForm submit={mockSubmit} />);
  });

  it("should display required error when value is invalid", async () => {
    fireEvent.submit(screen.getByRole("button"));

    expect(await screen.findAllByRole("alert")).toHaveLength(2);
    expect(mockSubmit).not.toBeCalled();
  });

  // TODO: uncomment this unite test when add the email address validation
  // it("should display matching error when email is invalid", async () => {
  //   fireEvent.input(screen.getByRole("textbox", { name: /Email address/i }), {
  //     target: {
  //       value: "test",
  //     },
  //   });

  //   fireEvent.input(screen.getByRole("textbox", { name: /message/i }), {
  //     target: {
  //       value: "New message to send to us",
  //     },
  //   });

  //   fireEvent.submit(screen.getByRole("button"));

  //   expect(await screen.findAllByRole("alert")).toHaveLength(1);
  //   expect(mockSubmit).not.toBeCalled();
  //   expect(screen.getByRole("textbox", { name: /Email address/i }).value).toBe(
  //     "test"
  //   );
  //   expect(screen.getByRole("textbox", { name: /message/i }).value).toBe(
  //     "New message to send to us"
  //   );
  // });

  it("should display min length error when message is short", async () => {
    fireEvent.input(screen.getByRole("textbox", { name: /Email address/i }), {
      target: {
        value: "test@mail.com",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /message/i }), {
      target: {
        value: "pass",
      },
    });

    fireEvent.submit(screen.getByRole("button"));

    expect(await screen.findAllByRole("alert")).toHaveLength(1);
    expect(mockSubmit).not.toBeCalled();
    expect(screen.getByRole("textbox", { name: /Email address/i }).value).toBe(
      "test@mail.com"
    );
    expect(screen.getByRole("textbox", { name: /message/i }).value).toBe(
      "pass"
    );
  });

  it("should not display error when value is valid", async () => {
    fireEvent.input(screen.getByRole("textbox", { name: /Full Name/i }), {
      target: {
        value: "Name Nami",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /Email address/i }), {
      target: {
        value: "test@mail.com",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /message/i }), {
      target: {
        value: "Long message to be sent",
      },
    });

    fireEvent.submit(screen.getByRole("button"));

    // expect(await screen.findAllByRole("alert")).toHaveLength(0);

    await waitFor(() => expect(screen.queryAllByRole("alert")).toHaveLength(0));
    expect(mockSubmit).toBeCalledWith({
      email: "test@mail.com",
      fullName: "Name Nami",
      messageContent: "Long message to be sent",
    });
    expect(screen.getByRole("textbox", { name: /Full Name/i }).value).toBe("");
    expect(screen.getByRole("textbox", { name: /Email address/i }).value).toBe(
      ""
    );
    expect(screen.getByRole("textbox", { name: /message/i }).value).toBe("");
  });
});
