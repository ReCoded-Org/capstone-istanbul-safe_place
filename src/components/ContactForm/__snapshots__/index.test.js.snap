// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaveMessageForm renders correctly and matches the snapshot 1`] = `
<div
  className="contactForm col-md-6 col-12"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="formInput"
    >
      <label
        htmlFor="fullName"
      >
        Full Name
        <input
          id="fullName"
          name="fullName"
          placeholder="Louis Li (optional)"
          type="text"
        />
      </label>
    </div>
    <div
      className="formInput"
    >
      <label
        htmlFor="email"
      >
        Email address
        <input
          id="email"
          name="email"
          placeholder="louis@example.com"
          type="email"
        />
      </label>
    </div>
    <div
      className="formInput"
    >
      <label
        htmlFor="messageContent"
      >
        Message
        <textarea
          aria-label="message"
          id="messageContent"
          name="messageContent"
          placeholder="type your message here ..."
        />
      </label>
    </div>
    <button
      className="submitBtn"
      type="submit"
    >
      Send
    </button>
  </form>
</div>
`;
